using_restricted_array.rb

```diff
@@ -7,41 +7,104 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  value = array[0]
+  until value == nil
+    value = array[i]
+    i += 1
+  end
+  return (i - 1)
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  value = array[0]
+  until value == nil
+    value = array[i]
+    puts value
+    i += 1
+  end
+  return value
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  value = array[0]
+  until value == nil
+    value = array[i]
+    if value == value_to_find
+      return true
+    end
+    i += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  max_value = array[0]
+  until array[i] == nil
+    if max_value < array[i]
+      max_value = array[i]
+    end
+    i += 1
+  end
+  return max_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  min_value = array[0]
+  until array[i] == nil
+    if min_value > array[i]
+      min_value = array[i]
+    end
+    i += 1
+  end
+  return min_value
 end
 
 # Reverses the values in the integer array in place
+
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+
+  while i < length/2
+    a = array[i]
+    b = array[length - 1 - i]
```

Comment:
You are using two temporary variables here, a and b. You could optimize the swap by just using one temporary variable like so:
`a = array[i]`
`array[i] = array[length -1-i]`
`array[length-1-i] = a`